core/templates/dev/head/player/AnswerClassificationService.js

```diff
@@ -0,0 +1,142 @@
+   * <ul>
+   */
+  var classifyAnswer = function(
+      answer, answerGroups, defaultOutcome, interactionRulesService) {
+    // Find the first group that contains a rule which returns true
+    for (var i = 0; i < answerGroups.length; i++) {
+      for (var j = 0; j < answerGroups[i].rule_specs.length; j++) {
+        var ruleSpec = answerGroups[i].rule_specs[j];
+        if (interactionRulesService[ruleSpec.rule_type](answer, ruleSpec.inputs)) {
+          return {
+            'outcome': answerGroups[i].outcome,
+            'answerGroupIndex': i,
+            'ruleSpecIndex': j
+          };
+        }
+      }
+    }
+
+    // If no rule in any answer group returns true, the default 'group' is
+    // returned. Throws an error if the default outcome is not defined.
+    if (defaultOutcome) {
+      return {
+        'outcome': defaultOutcome,
+        'answerGroupIndex': answerGroups.length,
+        'ruleSpecIndex': 0
+      };
+    } else {
+      warningsData.addWarning('No default outcome found.');
+    }
+  };
+
+  return {
+    /**
+     * Gets a promise to the matching answer group.
+     *
+     * @param {string} explorationId The exploration ID.
+     * @param {*} answer The answer that the user has submitted.
+     * @param {object} oldState The state where the user submitted the answer.
+     * @param {?function} interactionRulesService The service which contains the
+     *     rules of that interaction. If this is undefined, then the function
+     *     server-side classification.
+     *
+     * @return {promise} An promise for an object representing the answer group
+     *     with the following properties:
+     * <ul>
+     *   <li> **outcome**: the outcome of the answer group
+     *   <li> **answerGroupIndex**: the index of the matched answer group
+     *   <li> **ruleSpecIndex**: the index of the rule in the matched answer group
+     * <ul>
+     */
+    getMatchingClassificationResult: function(
+        explorationId, oldState, answer, interactionRulesService) {
+      var deferred = $q.defer();
+      if (interactionRulesService) {
+        var answerGroups = oldState.interaction.answer_groups;
+        var defaultOutcome = oldState.interaction.default_outcome;
+
+        deferred.resolve(classifyAnswer(
+          answer, answerGroups, defaultOutcome, interactionRulesService));
+      } else {
+        var classifyUrl = '/explorehandler/classify/' + explorationId;
+        $http.post(classifyUrl, {
+          old_state: oldState,
+          params: learnerParamsService.getAllParams(),
+          answer: answer
+        }).success(function(result) {
+          deferred.resolve({
+            outcome: result.outcome,
+            ruleSpecIndex: result.rule_spec_index,
+            answerGroupIndex: result.answer_group_index
+          });
+        });
+      }
+      return deferred.promise;
+    },
+    getMatchingEditorClassificationResult: function(
```

Comment:
It feels to me like there's a lot of duplication between this and the previous method. Why can't these be a single method?